library/std/unstable_state_preparation.qs

```diff
@@ -0,0 +1,459 @@
+        }
+        // target is now in |0> state up to the phase given by arg of newCoefficients.
+
+        // Continue recursion while there are control qubits.
+        if (IsRangeEmpty(rngControl)) {
+            let (abs, arg) = newCoefficients[0]!;
+            if (AbsD(arg) > tolerance) {
+                Exp([PauliI], -1.0 * arg, [register[idxTarget]]);
+            }
+        } elif (AnyOutsideToleranceCP(tolerance, newCoefficients)) {
+            let newControl = (RangeStart(rngControl) + 1)..RangeStep(rngControl)..RangeEnd(rngControl);
+            let newTarget = RangeStart(rngControl);
+            ApproximatelyUnprepareArbitraryState(tolerance, newCoefficients, newControl, newTarget, register);
+        }
+    }
+
+    /// # Summary
+    /// Applies a Pauli rotation conditioned on an array of qubits, truncating
+    /// small rotation angles according to a given tolerance.
+    ///
+    /// # Description
+    /// This applies a multiply controlled unitary operation that performs
+    /// rotations by angle $\theta_j$ about single-qubit Pauli operator $P$
+    /// when controlled by the $n$-qubit number state $\ket{j}$.
+    /// In particular, the action of this operation is represented by the
+    /// unitary
+    ///
+    /// $$
+    /// \begin{align}
+    ///     U = \sum^{2^n - 1}_{j=0} \ket{j}\bra{j} \otimes e^{i P \theta_j}.
+    /// \end{align}
+    /// ##
+    ///
+    /// # Input
+    /// ## tolerance
+    /// A tolerance below which small coefficients are truncated.
+    ///
+    /// ## coefficients
+    /// Array of up to $2^n$ coefficients $\theta_j$. The $j$th coefficient
+    /// indexes the number state $\ket{j}$ encoded in little-endian format.
+    ///
+    /// ## pauli
+    /// Pauli operator $P$ that determines axis of rotation.
+    ///
+    /// ## control
+    /// $n$-qubit control register that encodes number states $\ket{j}$ in
+    /// little-endian format.
+    ///
+    /// ## target
+    /// Single qubit register that is rotated by $e^{i P \theta_j}$.
+    ///
+    /// # Remarks
+    /// `coefficients` will be padded with elements $\theta_j = 0.0$ if
+    /// fewer than $2^n$ are specified.
+    internal operation ApproximatelyMultiplexPauli(
+        tolerance : Double,
+        coefficients : Double[],
+        pauli : Pauli,
+        control : Qubit[],
+        target : Qubit) : Unit is Adj + Ctl {
+
+        if pauli == PauliZ {
+            ApproximatelyMultiplexZ(tolerance, coefficients, control, target);
+        } elif pauli == PauliX {
+            within {
+                H(target);
+            } apply {
+                ApproximatelyMultiplexPauli(tolerance, coefficients, PauliZ, control, target);
+            }
+        } elif pauli == PauliY {
+            within {
+                Adjoint S(target);
+            } apply {
+                ApproximatelyMultiplexPauli(tolerance, coefficients, PauliX, control, target);
+            }
```

Comment:
This can be made more concise with `M.Q.Intrinsic.MapPauli`.

Reply:
Postponed for now. Filed https://github.com/microsoft/qsharp/issues/1104